### PR TITLE
Fix Sessions support in WooCommerce Analytics

### PR DESCRIPTION
**Pull request by puntope**
Add the next fixes and added functionality:

- https://github.com/Automattic/jetpack/pull/41085
- https://github.com/Automattic/jetpack/pull/41476
- https://github.com/Automattic/jetpack/pull/41754 ⛔ (Reverted)

## Proposed changes:

* Fix and tweak session_started event
* Add ClickHouse tracking support 
* Remove `woocommerceanalytics_` prefix when tracking to click house 

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Does this pull request change what data or activity we track or use?

YES -- https://github.com/Automattic/jpop-quill/issues/404

## Testing instructions:

Session Started Events:

* Set up a Jurassic Ninja site using the code in this PR for WooCommerce Analytics
* Verify the cache by hitting different pages with curl -I 
* When running it the second time you should see something like:

Screenshot 2025-01-15 at 13 50 25

See how it was before when the set cookie was set in PHP headers

Screenshot 2025-01-14 at 17 03 59

* Close all the incognito windows and open a new one, navigate to a Product (or perform any other event like going to checkout or cart)
* Verify you dont need to perform setup instructions in https://github.com/Automattic/analytics/pull/76 to test this PR as the library should be updated.
* See session_started event is triggered and also the event contains session_id and loads in both t.gif and w.gif in all the available events with `woocommerceanalytics_` prefix.
```
'woocommerceanalytics_session_started',
			'woocommerceanalytics_product_view',
			'woocommerceanalytics_cart_view',
			'woocommerceanalytics_add_to_cart',
			'woocommerceanalytics_remove_from_cart',
			'woocommerceanalytics_checkout_view',
			'woocommerceanalytics_product_checkout',
			'woocommerceanalytics_product_purchase',
			'woocommerceanalytics_order_confirmation_view',
			'woocommerceanalytics_search',

```
* Tests edge cases, like Empty carts or going to checkout when cart is empty or remove session cookie in between.

### Additional notes:

- Notice that the branch name is 0.42 but actually, we release it to 0.5.0 as commented [here](https://github.com/Automattic/jetpack/pull/41476#discussion_r1939134168)

### CODE REVIEW

**Review by ianlin — COMMENTED** [review]
Thanks for wrapping everything for a new release. Most of the scenarios worked expectedly. I will list only the results that might need further discussion.

### Setup

- Create a new JN site with latest WC 9.7.0
- Build jetpack using this branch by running `jetpack build plugins/jetpack`
- Copy the plugin to the site by running: `jetpack rsync jetpack profoundly-additional.jurassic.ninja@ssh.atomicsites.net:/srv/htdocs/wp-content/plugins/jetpack`

### Test cases that need discussion

#### 1. When the cart has products, remove session cookie and enter checkout page directly

This is kind of an edge case.

https://github.com/user-attachments/assets/dacf350a-7b18-44d8-91ed-9048e91da0ca

1. Add a product to cart
2. Remove the session cookie
3. Go to the checkout page directly
4. The session cookie is not created, the `session_started`, `checkout_view`, and `product_checkout` events are not triggered.

#### 2. Logged in user enters My Account page actually loads `w.gif`, and the `session started` event is triggered

https://github.com/user-attachments/assets/1dd7f2cb-e546-42a3-a3da-49e09292f2f5

1. Logged in to the store
2. Remove the session cookie
3. Go to My Account page
4. See the `w.gif` is actually loaded from the Network tab
5. See the `session_started` event is actually triggered

This behaviour is different from that in the PR description:

> Perform some not CH events like My Account View and verify only t.gif is loaded with the woocommerceanalytics_ prefix on the event name.

But I guess this is also acceptable, so I think maybe we can keep this behaviour?

#### 3. Cache miss on the single product

https://github.com/user-attachments/assets/a588739f-50c2-412d-8c98-b9a2fa36f6c0

1. Run `curl -I https://profoundly-additional.jurassic.ninja/shop/single` multiple times
2. See the cache is still `Miss` from the `x-ac` header
3. Run `curl -I https://profoundly-additional.jurassic.ninja` multiple times
4. See the cache is `Hit` from the `x-ac` header

Just to be clear that the `set-cookie` header is not there anymore. But the cache still shows `Miss` on the single product page.

However I don't think this is related to this PR, because when I use another brand new JN site **_without_** using this PR, the same cache miss behaviour still happen. I remember last time I tested it the cache is `Hit` on the single product page after the first call. Not sure if this is the correct way to verify the cache, and should we ask someone else that is more familiar with the cache topic?

**Comment by puntope** [comment]
> 2. Logged in user enters My Account page actually loads w.gif, and the session started event is triggered

This is ok. As `session_started` event is triggered in w.gif. Notice `myaccount_page_view` doesn't

**Comment by ianlin** [comment]
> Cache miss on the single product

Ah, I think I had an epic fail. I was testing it with the endpoint `/shop/single` and it redirects to `/product/single` so the cache MISS is expected. Testing using `/product/single` works as expected. Sorry for misleading.

**Comment by puntope** [comment]
> 1. When the cart has products, remove session cookie and enter checkout page directly

Interesting edge case. Took a while for me to understand whats going on here.

1. Checkout has 2 events recorded actually in the same request. `product_checkout` and `checkout_view`. 
2. In the first event. We check if cookie `woocommerceanalytics_session` is set.
3. Since the cookie was removed we add the JS code for adding the cookie + creating the session. Notice this code is not yet executed. Just queued.
4. In the second event under same request we check if cookie `woocommerceanalytics_session` is set.
5. Since the cookie was removed we add the JS code for adding the cookie + creating the session. Notice this code is not yet executed. Just queued.
6. We print the queued JS
7. There is a syntax error as const sesssionData was created twice and hence the code is not executed.

Screenshot 2025-02-25 at 13 49 30

This only happens in those events that include 2 record_event() calls in the same request. So checkout and order completed.

The solution was simple tho. Just checking if session_id is set. As all this is happening in the same request.

Added here --> [6cdfaa3](https://github.com/Automattic/jetpack/pull/41984/commits/6cdfaa395a074b6b1f54384027e491d06002dd7b)

**Comment by puntope** [comment]
> > Cache miss on the single product
> 
> Ah, I think I had an epic fail. I was testing it with the endpoint `/shop/single` and it redirects to `/product/single` so the cache MISS is expected. Testing using `/product/single` works as expected. Sorry for misleading.

No problem. Appreciated you checked twice. Better to be sure all is good.

**Comment by puntope** [comment]
@ianlin Thanks a bunch again for the review. And the bug you found. 

Let me know if everything works now.

**Comment by ianlin** [comment]
Thanks for the fix, it works now.

I found another edge case in My Account page that seems also related to the `sessionData` redeclare error you mentioned above.

https://github.com/user-attachments/assets/bb8597fc-514f-44e3-ae3d-6faaf461e1ed

1. Perform any actions that will not fire an event immediately until the next page load, such as
    - Add a product to cart
    - Remove a product from cart
2. Remove session cookie
3. Go to **My Account** page
4. No events are triggered, and the error `Uncaught SyntaxError: Identifier 'sessionData' has already been declared (at my-account/:585:19)` display in the dev console.

The error seems to only happens in My Account page, if I go to other pages like Cart or Checkout page, the `session_started` event will be triggered, as well as the event for the previous action before the session cookie was removed. However, the session ID for the event of the previous action was assigned to a new session ID, not sure if this scenario makes sense from the data analysis point of view?

For example:

- User adds a product to cart (session ID = abc123)
- For some reason the session cookie gets removed
- User goes to cart or checkout page
- The `session_started` and the previous `add_to_cart` event were sent (session ID = abc456)
- However, `add_to_cart` was actually belongs to the session abc123.

**Comment by puntope** [comment]
Thanks again @ianlin for those catches. I adapted again the code to add session_id and landing_page in the loop events added in sessions.

⚠️  I noticed that "Add to cart" in Product List when there are more items in cart (so cart quantity update) doesn't work. This seems to be another bug related to the package that I would solve in future PRs

File: projects/packages/woocommerce-analytics/src/class-woo-analytics-trait.php [hunk 1]
```diff
@@ -697,4 +699,35 @@ public function get_cart_items_count() {
...
+
+	/**
+	 * Return the landing page.
+	 * First try to get it from the cookie. Fallback to $this->landing_page.
+	 *
+	 * @return string
+	 */
+	public function get_landing_page() {
+		$cookie = $this->get_session_cookie();
+		return $cookie['landing_page'] ?? $this->session_id;
```
Comment (ianlin) [inline projects/packages/woocommerce-analytics/src/class-woo-analytics-trait.php]: I think it should be `$cookie['landing_page'] ?? $this->landing_page;`.

File: projects/packages/woocommerce-analytics/src/class-my-account.php [hunk 2]
```diff
@@ -65,11 +65,11 @@ public function track_tabs() {
 				if ( isset( $core_endpoints['view-order'] ) && $core_endpoints['view-order'] === $key && is_numeric( $value ) ) {
 					$initiator = get_query_var( '_wca_initiator' );
 					if ( 'number' === $initiator ) {
-						$this->record_event( 'woocommerceanalytics_my_account_order_number_click' );
+						$this->record_event( 'woocommerceanalytics_my_account_order_number_click', array(), null, false );
```
Comment (nerrad) [inline projects/packages/woocommerce-analytics/src/class-my-account.php]: I see you are assigning values explicitly for all the arguments because of the new `$clickhouse` parameter added to the `record_event` function. So the big change here is:

- Calls to `record_event` go to Clickhouse _by default_.
- Any events that previously went to Nosara that are _not_ updated to explicitly send false will end start going to Clickhouse instead of Nosara. Which is (I'm guessing) why you are explicitly setting `false` here.

At the risk of being a bit naive about the changes, I have a few questions:
- What rationale exists for introducing new default behaviour here to go to Clickhouse?
- What risk exists for any database gaps in existing code that calls `record_event` and the tracks are going to Clickhouse instead of Nosara? Have you included consumers of this package in your review of impact?

File: projects/packages/woocommerce-analytics/src/class-woo-analytics-trait.php [hunk 3]
```diff
@@ -410,7 +435,12 @@ public function process_event_properties( $event_name, $properties = array(), $p
 			)
 		);
 
-		$js = "{'_en': '" . esc_js( $event_name ) . "'";
+		// ch param is needed to identify ClickHouse queries in the JS Analytics library. Remove prefix also for ClickHouse
+		if ( $clickhouse ) {
+			$js = "{'_en': '" . esc_js( str_replace( 'woocommerceanalytics_', '', $event_name ) ) . "','ch':'1'";
```
Comment (nerrad) [inline projects/packages/woocommerce-analytics/src/class-woo-analytics-trait.php]: Why is the event prefix removed for ClickHouse?

**Review by ianlin — COMMENTED** [review]
Thanks for the fix, it was fixed. But a similar case still got the error `Uncaught SyntaxError: Identifier 'sessionData' has already been declared (at my-account/:585:19)`.

1. Remove session cookie
2. Perform any actions that will not fire an event immediately until the next page load, such as
    - Add a product to cart
    - Remove a product from cart 
3. Go to My Account page
4. No events are triggered, and the error `Uncaught SyntaxError: Identifier 'sessionData' has already been declared (at my-account/:585:19)` display in the dev console.

It works if the step 3 was not going to My Account page (for example cart / checkout pages)

Also there is a more extreme case:

1. Add a product to cart (session ID: abc123)
2. Remove session cookie
3. Remove a product from cart (no session ID)
4. Go to cart page
5. There are 5 events fired:
    - `session_started`: session ID abc456 ✅
    - `cart_view`: session ID abc456 ✅
    - `add_to_cart`: session ID abc123 ✅
    - `product_view`: session ID abc123 ✅
    - `remove_from_cart` session ID is neither abc123 nor abc456.

IMO the `remove_from_cart`'s session ID should be the same as `session_started` and `cart_view`, in this case `abc456`.

File: projects/packages/woocommerce-analytics/src/class-woo-analytics-trait.php (hunk 3 above)
Comment (puntope) [reply projects/packages/woocommerce-analytics/src/class-woo-analytics-trait.php]: Because all the events are populated in the ClickHouse table named `woocommerce_analytics_store_events` that is exclusive for WooCommerce Analytics. 

Since we found it redundant we decided to have them without prefix. Notice this is different that the Nosara tracks which includes more than just the `woocommerceanalytics_` events.

Do you suggest to keep it?

File: projects/packages/woocommerce-analytics/src/class-my-account.php (hunk 2 above)
Comment (puntope) [reply projects/packages/woocommerce-analytics/src/class-my-account.php]: > Calls to record_event go to Clickhouse by default.

Correct

> Any events that previously went to Nosara that are not updated to explicitly send false will end start going to Clickhouse instead of Nosara. Which is (I'm guessing) why you are explicitly setting false here.

Not exactly. The exact thing that we are doing is:

Any events that previously went to Nosara that are not updated to explicitly send false will start going to Clickhouse additionally to Nosara. 

> What rationale exists for introducing new default behaviour here to go to Clickhouse?

In the long/mid term we want to include all the events in the package to go to ClickHouse. For the first deliverable I excluded My Account events that are the only ones being sent exclusively to Nosara. Thats why I decided to keep the default as send both Nosara and ClickHouse. 

 > What risk exists for any database gaps in existing code that calls record_event and the tracks are going to Clickhouse instead of Nosara? Have you included consumers of this package in your review of impact?

Zero risk. All the events are being sent to Nosara (t.gif) with the `woocommerceanalytics_` prefix  even if we send it to ClickHouse.

File: projects/packages/woocommerce-analytics/src/class-woo-analytics-trait.php (hunk 1 above)
Comment (puntope) [reply projects/packages/woocommerce-analytics/src/class-woo-analytics-trait.php]: Thanks. Missed that one

File: projects/packages/woocommerce-analytics/src/class-my-account.php (hunk 2 above)
Comment (puntope) [reply projects/packages/woocommerce-analytics/src/class-my-account.php]: @nerrad thanks for your comments.

Maybe a better idea would be to have a constant listing the events to be send to CH instead. 

```
const CLICKHOUSE_EVENTS = [
 'woocommerceanalytics_add_to_cart',  
 'woocommerceanalytics_product_view',
 ... etc  
]
```

And then check if name of the event is in the array. So we can avoid the param and potential mistakes.

WDYT?

Comment (puntope) [reply projects/packages/woocommerce-analytics/src/class-my-account.php]: I added the refactor here. I thi nk is cleaner that way https://github.com/Automattic/jetpack/pull/41984/commits/8f7df4a780827e385af07a39487658f46b454479

File: projects/packages/woocommerce-analytics/src/class-woo-analytics-trait.php (hunk 3 above)
Comment (puntope) [reply projects/packages/woocommerce-analytics/src/class-woo-analytics-trait.php]: I keep the prefix here as well. So we can keep it homogenous with Nosara Tracks https://github.com/Automattic/jetpack/pull/41984/commits/fef8e05c4734e3117036d8978710cb3afcd16446

**Comment by puntope** [comment]
> Thanks for the fix, it was fixed. But a similar case still got the error Uncaught SyntaxError: Identifier 'sessionData' has already been declared (at my-account/:585:19).

I decided to remove sessions for not clickhouse events (my account) We need to study how my account logic works and we can fix those cases when we add them to clickhouse. 

WDYT?

**Comment by puntope** [comment]
> Also there is a more extreme case

Such an edge case :) 

This happens because Remove from Cart works with the API. It creates its own Session ID but the JS or generating the cookie is never rendered because it's an API call. And hence the cookie is never saved.

In the next request, when going to Cart Page (or any other) The cookie is not there and it's also a new request so the `session_id` param from the Remove from Cart request is not there either. As a result the new session is created.

So in summary what is happening is:

- 1 Session for the Add to Cart
The user removes that session manually from the inspector
- 1 Session for Remove from Cart API call
- 1 Session for Cart View  

Not an easy fix if we only count with JS cookies for preserving cache :( 

Maybe we can use WC->session and try to save there the session.  Unfortunately, the different hooks used in the package for triggering the events can be triggered before/after the session is available so not sure how many of the edge cases we will cover. 

I can log an issue for this case together with the Update Cart one and work on it in future releases. 

WDYT?

File: projects/packages/woocommerce-analytics/src/class-woo-analytics-trait.php [hunk 4]
```diff
@@ -434,15 +433,15 @@ public function process_event_properties( $event_name, $properties = array(), $p
...
 		);
 
+		$js = "{'_en': '" . esc_js( $event_name ) . "'";
+
 		// ch param is needed to identify ClickHouse queries in the JS Analytics library.
 		if ( $this->is_clickhouse( $event_name ) ) {
-			$js = "{'_en': '" . esc_js( $this->get_clickhouse_event( $event_name ) ) . "','ch':'1'";
-		} else {
-			$js = "{'_en': '" . esc_js( $event_name ) . "'";
+			$js .= "','ch':'1'";
```
Comment (ianlin) [inline projects/packages/woocommerce-analytics/src/class-woo-analytics-trait.php]: There an extra single quote `'` here that causes the syntax error:

```
_wca.push({'_en': 'woocommerceanalytics_session_started'','ch':'1','session_id': .....`
```

In this example there's an extra `'` after `woocommerceanalytics_session_started`.

**Comment by ianlin** [comment]
> I decided to remove sessions for not clickhouse events (my account) We need to study how my account logic works and we can fix those cases when we add them to clickhouse.
>
> WDYT?

Sounds good to me. I think the initial implementation was the same - entering My Account page wouldn't trigger `session_started` event if the session doesn't exist.

Just to make sure that it hasn't been implemented in this PR, right? Because I can still see `session_started` being triggered when entering My Account page.

**Comment by ianlin** [comment]
> I can log an issue for this case together with the Update Cart one and work on it in future releases.

Yeah that makes 100% sense since it's an extreme case.

**Comment by puntope** [comment]
Howdy @ianlin Sorry I forgot to commit my last changes. Now they are in the PR https://github.com/Automattic/jetpack/pull/41984/commits/c756d51791208b542b18edef8282094260027263.

Can you check again? thanks.

**Comment by ianlin** [comment]
Hey @puntope thanks for the changes. Found another case that is similar to the extreme case above, but this time I did not remove the session cookie manually, so I guess it's not an edge case.

1. Open a new incognito page, no session cookie being set so far
2. Go to shop page directly `https://dreamily-jolly-mantis.jurassic.ninja/shop`
3. The shop page does not trigger any CH events
4. Add some products to the cart
5. Open the mini cart in the right side bar
6. Remove an item from the mini cart
7. Go to a product page or the cart page
8. Notice the session ID of the event `woocommerceanalytics_remove_from_cart` is different from other events


https://github.com/user-attachments/assets/72efcd09-48d2-4f60-9a38-9f9b31a9db92

What I'm a bit curious about is: both `add_to_cart` and `remove_from_cart` actions will not send the event immediately since they are API calls, in the next page load the `add_to_cart` event has the same session ID with other events, but `remove_from_cart` does not?

**Comment by ianlin** [comment]
If this is not an easy fix then as you said we can log an issue and revisit it in the future, so we can get release most of the working analytics tracks.

**Comment by puntope** [comment]
> If this is not an easy fix then as you said we can log an issue and revisit it in the future, so we can get release most of the working analytics tracks.

@ianlin  What I did in the end was to create a new branch `release/woocommerce-analytics-5.0.0` we could merge for now this one here with all the fixes and I may implement the `page_view` and other bug fixes in a different PR and start a new fresh and clean release.

WDYT?

File: projects/packages/woocommerce-analytics/changelog/tweak-remov-prefix-woocommerceanalytics-clickhouse [hunk 5]
```diff
@@ -0,0 +1,4 @@
+Significance: patch
+Type: changed
+
+Remove prefix _woocommerceanalytics in ClickHouse events
```
Comment (ianlin) [inline projects/packages/woocommerce-analytics/changelog/tweak-remov-prefix-woocommerceanalytics-clickhouse]: I guess we don't need this change log as we are keeping the prefix for now. In the next PR maybe we can merge all change logs into one or two to make it more specifically before merging into trunk.

**Review by ianlin — APPROVED** [review]
Thanks for all the fixes. I tested again and all the ClickHouse events are triggered correctly except for the edge cases that we agree to move it to the next PR. Since there will be another PR before the actual release, can we update the PR title / description to make it clearer?
